schema/gen/go/genMap.go

```diff
@@ -0,0 +1,631 @@
+	// Perhaps surprisingly, the way to get the representation node pointer
+	//  does not actually depend on what the representation strategy is.
+	// REVIEW: this appears to be standard even across kinds; can we extract it?
+	doTemplate(`
+		func (n {{ .Type | TypeSymbol }}) Representation() ipld.Node {
+			return (*_{{ .Type | TypeSymbol }}__Repr)(n)
+		}
+	`, w, g.AdjCfg, g)
+}
+
+// --- Node interface satisfaction --->
+
+func (g mapGenerator) EmitNodeType(w io.Writer) {
+	// No additional types needed.  Methods all attach to the native type.
+}
+
+func (g mapGenerator) EmitNodeTypeAssertions(w io.Writer) {
+	doTemplate(`
+		var _ ipld.Node = ({{ .Type | TypeSymbol }})(&_{{ .Type | TypeSymbol }}{})
+		var _ schema.TypedNode = ({{ .Type | TypeSymbol }})(&_{{ .Type | TypeSymbol }}{})
+	`, w, g.AdjCfg, g)
+}
+
+func (g mapGenerator) EmitNodeMethodLookupString(w io.Writer) {
+	// What should be coercible in which directions (and how surprising that is) is an interesting question.
+	//  Most of the answer comes from considering what needs to be possible when working with PathSegment:
+	//   we *must* be able to accept a string in a PathSegment and be able to use it to navigate a map -- even if the map has complex keys.
+	//   For that to work out, it means if the key type doesn't have a string type kind, we must be willing to reach into its representation and use the fromString there.
+	//  If the key type *does* have a string kind at the type level, we'll use that; no need to consider going through the representation.
+	doTemplate(`
+		func (n {{ .Type | TypeSymbol }}) LookupString(k string) (ipld.Node, error) {
+			var k2 _{{ .Type.KeyType | TypeSymbol }}
+			{{- if eq .Type.KeyType.Kind.String "String" }}
+			if err := (_{{ .Type.KeyType | TypeSymbol }}__Style{}).fromString(&k2, k); err != nil {
+				return nil, err // TODO wrap in some kind of ErrInvalidKey
+			}
+			{{- else}}
+			if err := (_{{ .Type.KeyType | TypeSymbol }}__ReprStyle{}).fromString(&k2, k); err != nil {
+				return nil, err // TODO wrap in some kind of ErrInvalidKey
+			}
+			{{- end}}
+			v, exists := n.m[k2]
+			if !exists {
+				return ipld.Undef, ipld.ErrNotExists{ipld.PathSegmentOfString(k)}
+			}
+			{{- if .Type.ValueIsNullable }}
+			if v.m == schema.Maybe_Null {
+				return ipld.Null, nil
+			}
+			return {{ if not (MaybeUsesPtr .Type.ValueType) }}&{{end}}v.v, nil
+			{{- else}}
+			return v, nil
+			{{- end}}
+		}
+	`, w, g.AdjCfg, g)
+}
+
+func (g mapGenerator) EmitNodeMethodLookup(w io.Writer) {
+	// LookupNode will procede by cast if it can; or simply error if that doesn't work.
+	//  There's no attempt to turn the node (or its repr) into a string and then reify that into a key;
+	//   if you used a Node here, you should've meant it.
+	// REVIEW: by comparison structs will coerce anything stringish silently...!  so we should figure out if that inconsistency is acceptable, and at least document it if so.
```

Comment:
This is the second half of the questions in the area of https://github.com/ipld/go-ipld-prime/pull/52/files#r413877995 .